integration-tests/byoid/integration_test.go

```diff
@@ -0,0 +1,211 @@
+func TestMain(m *testing.M) {
+	keyFileName := os.Getenv(envCredentials)
+	if keyFileName == "" {
+		log.Fatalf("Please set %s to your keyfile", envCredentials)
+	}
+
+	projectID = os.Getenv(envProject)
+	if projectID == "" {
+		log.Fatalf("Please set %s to the ID of the project", envProject)
+	}
+
+	oidcAudience = os.Getenv(envAudienceOIDC)
+	if oidcAudience == "" {
+		log.Fatalf("Please set %s to the OIDC Audience", envAudienceOIDC)
+	}
+
+	kf, err := os.Open(keyFileName)
+	if err != nil {
+		log.Fatalf("Failed to open file '%s': %v", keyFileName, err)
+	}
+	defer kf.Close()
+
+	decoder := json.NewDecoder(kf)
+	var keyFileSettings keyFile
+	err = decoder.Decode(&keyFileSettings)
+	if err != nil {
+		log.Fatalf("Keyfile '%s' stored in improper format: %v", keyFileName, err)
+	}
+
+	clientID = fmt.Sprintf("projects/-/serviceAccounts/%s", keyFileSettings.ClientEmail)
+
+	iam, err := iamcredentials.NewService(context.Background())
+	if err != nil {
+		log.Fatalf("Unable to access iamcredentials service: %v", err)
+	}
+
+	idTokenInfo, err := iam.Projects.ServiceAccounts.GenerateIdToken(clientID, &iamcredentials.GenerateIdTokenRequest{
+		Audience:     oidcAudience,
+		IncludeEmail: true,
+	}).Do()
+	if err != nil {
+		log.Fatalf("Failed to generate ID Token: %v", err)
+	}
+
+	oidcToken = idTokenInfo.Token
+
+	// This line runs all of our individual tests
+	os.Exit(m.Run())
+}
+
+// testBYOID makes sure that the default credentials works for
+// whatever preconditions have been set beforehand
+// by using those credentials to run our client libraries.
+//
+// In each test we will set up whatever preconditions we need,
+// and then use this function.
+func testBYOID(t *testing.T, c config, env map[string]string) {
+	t.Helper()
+
+	// Set up config file.
+	configFile, err := ioutil.TempFile("", "config.json")
+	if err != nil {
+		t.Fatalf("Error creating config file: %v", err)
+	}
+	defer os.Remove(configFile.Name())
+
+	err = json.NewEncoder(configFile).Encode(c)
+	if err != nil {
+		t.Errorf("Error writing to config file: %v", err)
+	}
+	configFile.Close()
+
+	// Set up our environment variables.
+	if env == nil {
+		env = make(map[string]string)
+	}
+	env["GOOGLE_APPLICATION_CREDENTIALS"] = configFile.Name()
+	for key, value := range env {
+		oldValue := os.Getenv(key)
+		os.Setenv(key, value)
+		defer os.Setenv(key, oldValue)
+	}
+
+	// If our config file and environment variables were set up correctly,
+	// finding the default credentials should work correctly.
+	credentials, err := google.FindDefaultCredentials(context.Background(), "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		t.Fatalf("Could not find Default credentials: %v", err)
+	}
+
+	// Once the default credentials are obtained,
+	// we should be able to access Google Cloud resources.
+	client := oauth2.NewClient(context.Background(), credentials.TokenSource)
+	dnsService, err := dns.New(client)
```

Comment:
nit: Use NewService. Most of the prvious lines can be removed if you use the option package to pass in the location of the file. It will call google.FindDefaultCredentials under the hood.

Reply:
done